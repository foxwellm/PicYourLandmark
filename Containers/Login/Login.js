import React, { Component } from 'react';
import { StyleSheet, Text, View, TouchableOpacity, ImageBackground, TextInput } from 'react-native';
import { Icon } from 'react-native-elements';

export class Login extends Component {
  constructor() {
    super()
    this.state = {
      username: '',
      password: '',
      confirmPassword: '',
      email: 'new4@gmail.com',
      isLoginPage: true
    }
  }

  createAccount = async () => {
    const { username, password, confirmPassword, email } = this.state
    const url = `https://pic-landmark-api.herokuapp.com/api/v1/users/?email=${email}&username=${username}&password=${password}&password_confirmation=${confirmPassword}`
    const response = await fetch(url, { method: 'POST', headers: { 'Content-type': 'application/json' }})
    const result = await response.json()
    this.setState({ isLoginPage: true })
    this.props.setUserLoginId(Date.now())
  }

  toggleLoginPage = () => {
    this.setState({ isLoginPage: !this.state.isLoginPage })
  }

  render() {
    if (this.props.currentUserId >= 0) {
      return (
        <View style={styles.container}>
          <ImageBackground source={require('../../assets/goldengate.jpg')} style={styles.imageBackground}>
            <View style={styles.overlay} />
            <Icon color="white" name="user-circle" type="font-awesome" size={70} padding={15} top={50} />
              <TouchableOpacity style={[styles.button, { backgroundColor: '#e9e9e9', top: -10 }]} onPress={() => this.props.setUserLoginId(-1)}>
                <Text style={{ textAlign: 'center' }}>Log out</Text>
              </TouchableOpacity>
          </ImageBackground>
        </View>
      )
    } else if (this.state.isLoginPage) {
      return (
        <View style={styles.container}>
          <ImageBackground source={require('../../assets/goldengate.jpg')} style={styles.imageBackground}>
            <View style={styles.overlay} />
            <Icon color="white" name="user-circle" type="font-awesome" size={70} padding={15} top={50} />
            <View>
              <TextInput
                placeholder='Username'
                style={styles.inputBox}
                onChangeText={(username) => this.setState({ username })}
                value={this.state.username}
              />
              <TextInput
                placeholder='Password'
                secureTextEntry={true}
                password={true}
                style={[styles.inputBox, { top: -20 }]}
                onChangeText={(password) => this.setState({ password })}
                value={this.state.password}
              />
              <TouchableOpacity style={[styles.button, { backgroundColor: '#e9e9e9', top: -10 }]}>
                <Text style={{ textAlign: 'center' }}>Log in</Text>
              </TouchableOpacity>
              <TouchableOpacity style={styles.button} onPress={() => this.toggleLoginPage()}>
                <Text style={[styles.buttonText, { color: 'white' }]}>Click here to create an account</Text>
              </TouchableOpacity>
            </View>
          </ImageBackground>
        </View>
      )
    } else {
      return (
        <View style={styles.container}>
          <ImageBackground source={require('../../assets/goldengate.jpg')} style={styles.imageBackground}>
            <View style={styles.overlay} />
            <Icon color="white" name="user-circle" type="font-awesome" size={70} padding={15} top={40} />
            <View>
              <TextInput
                placeholder='Username'
                style={styles.inputBox}
                onChangeText={(username) => this.setState({ username })}
                value={this.state.username}
              />
              <TextInput
                placeholder='Password'
                secureTextEntry={true}
                password={true}
                style={[styles.inputBox, { top: -20 }]}
                onChangeText={(password) => this.setState({ password })}
                value={this.state.password}
              />
              <TextInput
                placeholder='Confirm Password'
                secureTextEntry={true}
                password={true}
                style={[styles.inputBox, { top: -10 }]}
                onChangeText={(confirmPassword) => this.setState({ confirmPassword })}
                value={this.state.confirmPassword}
              />
              <TouchableOpacity style={[styles.button, { backgroundColor: '#e9e9e9', top: 5 }]} onPress={() => this.createAccount()}>
                <Text style={styles.buttonText}>Create account</Text>
              </TouchableOpacity>
              <TouchableOpacity style={styles.button} onPress={() => this.toggleLoginPage()}>
                <Text style={[styles.buttonText, { color: 'white', top: 5 }]}>Already a member? Click here to log in</Text>
              </TouchableOpacity>
            </View>
          </ImageBackground>
        </View>
      )
    }
  }
}

const styles = StyleSheet.create({
  container: {
    display: 'flex',
    flexDirection: 'column',
    height: '90%',
    width: '100%',
    alignItems: 'center'
  },
  overlay: {
    position: 'absolute',
    top: 0,
    right: 0,
    bottom: 0,
    left: 0,
    backgroundColor: 'black',
    opacity: 0.4,
    height: '100%',
  },
  imageBackground: {
    width: '100%',
    height: '100%',
    display: 'flex',
    flex: 1,
    flexDirection: 'column',
    justifyContent: 'space-evenly',
    alignItems: 'center'
  },
  inputBox: {
    height: 40,
    width: 200,
    borderColor: 'gray',
    borderWidth: 2,
    backgroundColor: 'white',
    opacity: 0.7,
    padding: 5,
    textAlign: 'center',
    top: -30
  },
  button: {
    width: 200,
    padding: 10,
    borderRadius: 15,
  },
  buttonText: {
    textAlign: 'center'
  }
});

export default Login;