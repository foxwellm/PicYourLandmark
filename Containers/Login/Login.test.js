import React from 'react';
import { Login } from './Login';
import { shallow } from 'enzyme';
import { TouchableOpacity } from 'react-native';
import renderer from 'react-test-renderer'

describe('Login', () => {
  let wrapper;
  let mockSetUserLogin;
  let mockFetchUserInfo;
  let mockChangeCurrrentPage;
  let mockSetLoggedOutMessage;
  let mockSetUserLogout

  beforeEach(() => {
    mockSetUserLogin = jest.fn()
    mockFetchUserInfo = jest.fn()
    mockChangeCurrrentPage = jest.fn()
    mockSetLoggedOutMessage = jest.fn()
    mockSetUserLogout = jest.fn()
    wrapper = shallow(<Login currentUserId={1}
      setUserLogin={mockSetUserLogin}
      changeCurrentPage={mockChangeCurrrentPage}
      fetchUserInfo={mockFetchUserInfo}
      setLoggedOutMessage={mockSetLoggedOutMessage}
      setUserLogout={mockSetUserLogout}
    />)
  });

  it('should have a proper default state', () => {
    const expected = {
      username: '',
      password: '',
      confirmPassword: '',
      email: '',
      isLoginPage: true
    }

    expect(wrapper.state()).toEqual(expected);
  });

  describe('Login rendering', () => {
    it('should match the snapshot when the currentUserId is greater than or equal to 0', () => {
      wrapper.setProps({ currentUserId: -1 })
      expect(wrapper).toMatchSnapshot();
      wrapper.setProps({ currentUserId: 1 })
      expect(wrapper).toMatchSnapshot();
    });

    it('should match the snapshot when this.state.isLoginPage is true', () => {
      expect(wrapper).toMatchSnapshot();
    });

    it('should match the snapshot when this.state.isLoginPage is false', () => {
      wrapper.setState({ isLoginPage: false })
      expect(wrapper).toMatchSnapshot();
    });
  });

  describe('Login buttons', () => {
    it('logout button should call setUserLogout when pressed', () => {
      wrapper.find(TouchableOpacity).at(0).simulate('press')
      expect(mockSetUserLogout).toHaveBeenCalled()
    });

    it('login button should call this.loginUser when pressed', () => {
      wrapper.instance().loginUser = jest.fn()
      wrapper.setProps({ currentUserId: -1 })
      wrapper.setState({ isLoginPage: true })
      wrapper.find(TouchableOpacity).at(0).simulate('press')
      expect(wrapper.instance().loginUser).toHaveBeenCalled()
    });

    it('Click here to create account button should call this.toggleLoginPage when pressed', () => {
      wrapper.instance().toggleLoginPage = jest.fn()
      wrapper.setProps({ currentUserId: -1 })
      wrapper.setState({ isLoginPage: true })
      wrapper.find(TouchableOpacity).at(1).simulate('press')
      expect(wrapper.instance().toggleLoginPage).toHaveBeenCalled()
    });

    it('Create account button should call this.createAccount when pressed', () => {
      wrapper.instance().createAccount = jest.fn()
      wrapper.setProps({ currentUserId: -1 })
      wrapper.setState({ isLoginPage: false })
      wrapper.find(TouchableOpacity).at(0).simulate('press')
      expect(wrapper.instance().createAccount).toHaveBeenCalled()
    });

    it('Already a member? Click here to log in button should call this.toggleLoginPage when pressed', () => {
      wrapper.instance().toggleLoginPage = jest.fn()
      wrapper.setProps({ currentUserId: -1 })
      wrapper.setState({ isLoginPage: false })
      wrapper.find(TouchableOpacity).at(1).simulate('press')
      expect(wrapper.instance().toggleLoginPage).toHaveBeenCalled()
    });
  });

  describe('createAccount method', () => {
    it.skip('createAccount method should make post request when createAccount is called', async () => {
      window.fetch = jest.fn()
      let email = 'tester01@gmail.com'
      let username = 'tester01'
      let password = 'abc'
      let confirmPassword = 'abc'
      let mockUrl = `https://pic-landmark-api.herokuapp.com/api/v1/users/?email=${email}&username=${username}&password=${password}&password_confirmation=${confirmPassword}`

      wrapper.setState({ email, username, password, confirmPassword })
      await wrapper.instance().createAccount()
      expect(window.fetch).toHaveBeenCalledWith(mockUrl, { method: 'POST', headers: { 'Content-type': 'application/json' } })
    });

    it('createAccount method should send back user id and username, and call setUserLogin and changeCurrentPage methods if everything is ok', async () => {
      window.fetch = jest.fn().mockImplementation(() => Promise.resolve({
        json: () => Promise.resolve({
          "id": 27,
          "email": "landmarker@example.com",
          "username": "landmarker",
          "profile_url": "",
          "locations": []
        }),
        status: 200,
        ok: true
      }));
      await wrapper.instance().createAccount();
      expect(mockSetUserLogin).toHaveBeenCalledWith(27, "landmarker");
    });

    it('createAccount method should throw error if everything is not ok', async () => {
      const expected = Error('Error fetching data');
      window.fetch = jest.fn().mockImplementation(() => Promise.resolve({
        ok: false,
        json: jest.fn().mockReturnValue('Error fetching data')
      }));
      expect(wrapper.instance().createAccount()).rejects.toEqual(expected)
    });
  });

  describe('loginUser method', () => {
    it('loginUser method should call fetchUserInfo when loginUser is called', () => {
      wrapper.instance().loginUser()
      expect(mockFetchUserInfo).toHaveBeenCalled()
    });
  })

  describe('toggleLoginPage method', () => {
    it('toggleLoginPage method should set state with isLoginPage value', () => {
      wrapper.instance().toggleLoginPage()
      expect(wrapper.state().isLoginPage).toEqual(false)
    });
  });

  describe('handleUsernameText method', () => {
    it('should update the username on text input on login page', () => {
      const instanceOf = renderer.create(<Login isLoginPage={true} currentUserId={-1} />).getInstance()
      instanceOf.handleUsernameText('username1')
      expect(instanceOf.state.username).toEqual('username1')
    });

    it('should update the username on text input on create new account page', () => {
      const instanceOf = renderer.create(<Login isLoginPage={false} currentUserId={-1} />).getInstance()
      instanceOf.handleUsernameText('username2')
      expect(instanceOf.state.username).toEqual('username2')
    });
  });

  describe('handlePasswordText method', () => {
    it('should update the password on text input on login page', () => {
      const instanceOf = renderer.create(<Login isLoginPage={true} currentUserId={-1} />).getInstance()
      instanceOf.handlePasswordText('password1')
      expect(instanceOf.state.password).toEqual('password1')
    });

    it('should update the password on text input on create new account page', () => {
      const instanceOf = renderer.create(<Login isLoginPage={false} currentUserId={-1} />).getInstance()
      instanceOf.handlePasswordText('password2')
      expect(instanceOf.state.password).toEqual('password2')
    });

    it('should update the email on text input on create new account page', () => {
      const instanceOf = renderer.create(<Login isLoginPage={false} currentUserId={-1} />).getInstance()
      instanceOf.handleEmailText('tester001@gmail.com')
      expect(instanceOf.state.email).toEqual('tester001@gmail.com')
    });
  });

  describe('handleConfirmPasswordText method', () => {
    it('should update the confirm password on text input on create new account page', () => {
      const instanceOf = renderer.create(<Login isLoginPage={false} currentUserId={-1} />).getInstance()
      instanceOf.handleConfirmPasswordText('confirmPassword1')
      expect(instanceOf.state.confirmPassword).toEqual('confirmPassword1')
    });
  });
});
