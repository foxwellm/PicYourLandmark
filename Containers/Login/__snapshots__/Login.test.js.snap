// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login renders correctly 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "display": "flex",
      "flexDirection": "column",
      "height": "90%",
      "width": "100%",
    }
  }
>
  <View
    accessibilityIgnoresInvertColors={true}
    style={
      Object {
        "alignItems": "center",
        "display": "flex",
        "flex": 1,
        "flexDirection": "column",
        "height": "100%",
        "justifyContent": "space-evenly",
        "width": "100%",
      }
    }
  >
    <Image
      source={1}
      style={
        Array [
          Object {
            "bottom": 0,
            "left": 0,
            "position": "absolute",
            "right": 0,
            "top": 0,
          },
          Object {
            "height": "100%",
            "width": "100%",
          },
          undefined,
        ]
      }
    />
    <View
      style={
        Object {
          "backgroundColor": "black",
          "bottom": 0,
          "height": "100%",
          "left": 0,
          "opacity": 0.4,
          "position": "absolute",
          "right": 0,
          "top": 0,
        }
      }
    />
    <View>
      <View
        padding={15}
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "transparent",
            "justifyContent": "center",
          }
        }
        theme={
          Object {
            "colors": Object {
              "disabled": "hsl(208, 8%, 90%)",
              "divider": "#bcbbc1",
              "error": "#ff190c",
              "grey0": "#393e42",
              "grey1": "#43484d",
              "grey2": "#5e6977",
              "grey3": "#86939e",
              "grey4": "#bdc6cf",
              "grey5": "#e1e8ee",
              "greyOutline": "#bbb",
              "platform": Object {
                "android": Object {
                  "error": "#f44336",
                  "primary": "#2196f3",
                  "secondary": "#9C27B0",
                  "success": "#4caf50",
                  "warning": "#ffeb3b",
                },
                "ios": Object {
                  "error": "#ff3b30",
                  "primary": "#007aff",
                  "secondary": "#5856d6",
                  "success": "#4cd964",
                  "warning": "#ffcc00",
                },
              },
              "primary": "#2089dc",
              "searchBg": "#303337",
              "secondary": "#8F0CE8",
              "success": "#52c41a",
              "warning": "#faad14",
            },
          }
        }
        top={50}
        underlayColor="white"
      >
        <Text />
      </View>
    </View>
    <View>
      <TextInput
        allowFontScaling={true}
        onChangeText={[Function]}
        placeholder="Username"
        style={
          Object {
            "backgroundColor": "white",
            "borderColor": "gray",
            "borderWidth": 2,
            "height": 40,
            "opacity": 0.7,
            "padding": 5,
            "textAlign": "center",
            "top": -20,
            "width": 200,
          }
        }
        underlineColorAndroid="transparent"
        value=""
      />
      <TextInput
        allowFontScaling={true}
        onChangeText={[Function]}
        password={true}
        placeholder="Password"
        secureTextEntry={true}
        style={
          Object {
            "backgroundColor": "white",
            "borderColor": "gray",
            "borderWidth": 2,
            "height": 40,
            "opacity": 0.7,
            "padding": 5,
            "textAlign": "center",
            "top": -10,
            "width": 200,
          }
        }
        underlineColorAndroid="transparent"
        value=""
      />
      <TextInput
        allowFontScaling={true}
        onChangeText={[Function]}
        password={true}
        placeholder="Confirm Password"
        secureTextEntry={true}
        style={
          Object {
            "backgroundColor": "white",
            "borderColor": "gray",
            "borderWidth": 2,
            "height": 40,
            "opacity": 0.7,
            "padding": 5,
            "textAlign": "center",
            "width": 200,
          }
        }
        underlineColorAndroid="transparent"
        value=""
      />
      <View
        accessible={true}
        isTVSelectable={true}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "backgroundColor": "#e9e9e9",
            "borderRadius": 15,
            "opacity": 1,
            "padding": 10,
            "top": 10,
            "width": 200,
          }
        }
      >
        <Text>
          Create account
        </Text>
      </View>
    </View>
  </View>
</View>
`;
